ua_generate_*** CMake macros are broken when using the GNU Makefile CMake generator

## Description
The `ua_generate_***` CMake macros generate files in the `${UA_GEN_ID_OUTPUT_DIR}` directory, that by default is `${PROJECT_BINARY_DIR}/src_generated`. If this directory does not already exists (as it is the case for a new build directory) the generation fails when using the GNU Makefile CMake generator, as the `${PROJECT_BINARY_DIR}/src_generated` is not automatically generated by the build system. The exact error message is the following:
~~~
[ 20%] Generating src_generated/example_nodeids.h
Traceback (most recent call last):
  File "/home/straversaro/src/open62541/build/install/share/open62541/tools/generate_nodeid_header.py", line 24, in <module>
    fh = open(args.outfile + ".h", "wt", encoding='utf8')
IOError: [Errno 2] No such file or directory: '/home/straversaro/src/open62541-issue/build/src_generated/example_nodeids.h'
CMakeFiles/server_nodeset.dir/build.make:77: recipe for target 'src_generated/example_nodeids.h' failed
make[2]: *** [src_generated/example_nodeids.h] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/server_nodeset.dir/all' failed
make[1]: *** [CMakeFiles/server_nodeset.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
~~~

This is problem does not apply on `Ninja`, because in Ninja:
> Output directories are always implicitly created before running the command that relies on them.
see https://ninja-build.org/manual.html#_comparison_to_make 

## Background Information / Reproduction Steps
On a Ubuntu 18.04 terminal in which open62541 is installed: 
~~~
git clone https://gist.github.com/traversaro/66629fa5b7cb3ecc2f9aa6f9a390a448 open62541-issue
cd open62541-issue
mkdir build
cd build
ccmake ..
make
~~~

## Checklist

Please provide the following information:

 - [x] open62541 Version (release number or git tag): 2b4c01f71fe5045bd5c7de4afdcc2c01e4acdeae
 - [x] Other OPC UA SDKs used (client or server): Not Applicable
 - [x] Operating system: Ubuntu 18.04 
 - [x] Logs (with `UA_LOGLEVEL` set as low as necessary) attached: Not Applicable
 - [x] Wireshark network dump attached: Not Applicable 
 - [x] Self-contained code example attached
 - [x] Critical issue

--- a/tools/cmake/macros_public.cmake
+++ b/tools/cmake/macros_public.cmake
@@ -43,6 +43,11 @@ function(ua_generate_nodeid_header)
     add_custom_target(${UA_GEN_ID_TARGET_PREFIX}-${UA_GEN_ID_TARGET_SUFFIX} DEPENDS
         ${UA_GEN_ID_OUTPUT_DIR}/${UA_GEN_ID_NAME}.h
     )
+    
+    # Make sure that the output directory exists 
+    if(NOT EXISTS ${UA_GEN_ID_OUTPUT_DIR})
+        file(MAKE_DIRECTORY ${UA_GEN_ID_OUTPUT_DIR})
+    endif()
 
     # Header containing defines for all NodeIds
     add_custom_command(OUTPUT ${UA_GEN_ID_OUTPUT_DIR}/${UA_GEN_ID_NAME}.h
@@ -148,6 +153,11 @@ function(ua_generate_datatypes)
     foreach(f ${UA_GEN_DT_FILES_BSD})
         set(BSD_FILES_TMP ${BSD_FILES_TMP} "--type-bsd=${f}")
     endforeach()
+    
+    # Make sure that the output directory exists 
+    if(NOT EXISTS ${UA_GEN_DT_OUTPUT_DIR})
+        file(MAKE_DIRECTORY ${UA_GEN_DT_OUTPUT_DIR})
+    endif()
 
     add_custom_command(OUTPUT ${UA_GEN_DT_OUTPUT_DIR}/${UA_GEN_DT_NAME}_generated.c
         ${UA_GEN_DT_OUTPUT_DIR}/${UA_GEN_DT_NAME}_generated.h
@@ -293,6 +303,11 @@ function(ua_generate_nodeset)
     foreach(f ${UA_GEN_NS_FILE})
         set(FILE_LIST ${FILE_LIST} "--xml=${f}")
     endforeach()
+    
+    # Make sure that the output directory exists 
+    if(NOT EXISTS ${UA_GEN_NS_OUTPUT_DIR})
+        file(MAKE_DIRECTORY ${UA_GEN_NS_OUTPUT_DIR})
+    endif()
 
     add_custom_command(OUTPUT ${UA_GEN_NS_OUTPUT_DIR}/ua_namespace${FILE_SUFFIX}.c
                        ${UA_GEN_NS_OUTPUT_DIR}/ua_namespace${FILE_SUFFIX}.h

